test/mitmproxy/proxy/protocol/test_websocket.py

```diff
@@ -329,7 +331,12 @@ def test_pong(self):
 
         assert frame.header.opcode == websockets.OPCODE.PONG
         assert frame.payload == b'foobar'
-        assert self.master.has_log("Pong Received from server", "info")
+        for i in range(20):
+            if self.master.has_log("Pong Received from server", "info"):
+                break
+            time.sleep(0.01)
+        else:
+            raise AssertionError("No pong seen")
```

Comment:
Just to clarify:
This basically tries 20 times to look for a pong log message, and then fails if nothing is found?
So kinda like an implicit timeout / early exit if the test would fail?

Reply:
Yes. The test suite will hopefully become nicer as we progress here. This should really be a `wait_for_log` method in our built-in taddons mechanism. Much of the test suite should be migrated over to use that, and server variants should really be fixtures, etc. Our test suite is becoming a burden, to be honest - it took much, much longer to adapt the tests for this patch than it did to do the core migration.